lib/scanner/blocks.go

```diff
@@ -24,48 +24,64 @@ type Counter interface {
 	Update(bytes int64)
 }
 
-// Blocks returns the blockwise hash of the reader.
-func Blocks(ctx context.Context, r io.Reader, blocksize int, sizehint int64, counter Counter, useWeakHashes bool) ([]protocol.BlockInfo, error) {
-	hf := sha256.New()
-	hashLength := hf.Size()
-
-	var mhf io.Writer
-	var whf hash.Hash32
+func createWriterAndWeakHashFun(h hash.Hash, useWeakHashes bool) (io.Writer, hash.Hash32) {
+	var writer io.Writer
+	var h32 hash.Hash32
 
 	if useWeakHashes {
-		whf = adler32.New()
-		mhf = io.MultiWriter(hf, whf)
+		h32 = adler32.New()
+		writer = io.MultiWriter(h, h32)
 	} else {
-		whf = noopHash{}
-		mhf = hf
+		h32 = noopHash{}
+		writer = h
 	}
+	return writer, h32
+}
 
+func prepareReader(sizehint int64, reader io.Reader, blocksize int, hashLength int) ([]protocol.BlockInfo, []byte, *io.LimitedReader) {
 	var blocks []protocol.BlockInfo
-	var hashes, thisHash []byte
+	var hashes []byte
 
 	if sizehint >= 0 {
 		// Allocate contiguous blocks for the BlockInfo structures and their
 		// hashes once and for all, and stick to the specified size.
-		r = io.LimitReader(r, sizehint)
+		reader = io.LimitReader(reader, sizehint)
 		numBlocks := int(sizehint / int64(blocksize))
 		blocks = make([]protocol.BlockInfo, 0, numBlocks)
 		hashes = make([]byte, 0, hashLength*numBlocks)
 	}
 
+	lreader := io.LimitReader(reader, int64(blocksize)).(*io.LimitedReader)
+	return blocks, hashes, lreader
+}
+
+// Blocks returns the blockwise hash of the reader.
+func Blocks(ctx context.Context, reader io.Reader, blocksize int, sizehint int64, counter Counter, useWeakHashes bool) ([]protocol.BlockInfo, error) {
+	if counter == nil {
+		counter = &NoopCounter{}
+	}
+
+	strongHash := sha256.New()
+	hashLength := strongHash.Size()
+
+	writer, weakHash := createWriterAndWeakHashFun(strongHash, useWeakHashes)
```

Comment:
Given function signatures in go start with func rather than fun, let's call it func here too.